File: classes/java.base/src/java/lang/Object.java
package java.lang;

public class Object {
	
	private static native void registerNatives();
	
	static {
		registerNatives();
	}
	
	@Deprecated(since="9")
	protected void finalized() throws Throwable{}
	
	public native int hashCode();
	
	public Object() {}
	
	protected native Object clone() throws CloneNotSupportedException;
	
	public final native Class<?> getClass();
	
	public String toString() {
		return getClass()+"@"+hashCode();
	}
	
	public boolean equals(Object o) {
		return this==o;
	}
	
	public final native void notify();
	public final native void notifyAll();
	public final native void wait(long milis) throws InterruptedException;
	public final void wait() throws InterruptedException {
		wait(0);
	}
	public final void wait(long milis,int nanos) throws InterruptedException {
		if(nanos>0)
			wait(milis+1);
		else
			wait(milis);
	}
	
	
}

File: natives/java/base/Object.lua
local Object = {};

local libjvm = require"java.jvm";

local hashCode;
local getClass;
local wait;
local notify;
local notifyAll;
local clone;


function Object.Java_java_lang_Object_registerNatives(env,cl)
  local natives = {
    hashCode=hashCode,
    getClass=getClass,
    wait0=wait,
    notify0=notify,
    notifyAll0=notifyAll,
    clone=clone
  }
  env.RegisterNatives(cl,natives);
end

local function hashCode(env,o)
  return libjvm.IdentityHashCode(libjvm.DerefObject(o));
end

local function getClass(env,o)
  return libjvm.toJNIObject(libjvm.ClassObject(libjvm.GetClass(libjvm.DerefObject(o))));
end

return Object;